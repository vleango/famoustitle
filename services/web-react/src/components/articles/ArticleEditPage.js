import React, { Component } from 'react';
import { connect } from 'react-redux';
import ReactMarkdown from 'react-markdown';
import { Button, Form, FormGroup, Input } from 'reactstrap';

import Header from '../shared/headers/Header';

export class ArticleEditPage extends Component {

  constructor(props) {
    super(props);
    this.state = {
      title: props.article.title,
      body: props.article.body
    };
  }

  onSubmitArticle = (e) => {
  }

  onInputChange = (e) => {
    const field = e.target.name;
    const value = e.target.value;
    this.setState(() => ({ [field]: value }));
  }

  render() {
    return (
      <div>
        <Header resourceTitle={this.props.article.title} />
        <div className="container">
            <Form onSubmit={this.onSubmitArticle}>
              <div className="row">
                <div className="col">
                  <FormGroup>
                    <Input type="text"
                      name="title"
                      value={this.state.title}
                      placeholder="Title"
                      onChange={this.onInputChange} />
                  </FormGroup>
                  <FormGroup>
                    <Input type="textarea"
                      rows="20"
                      name="body"
                      value={this.state.body}
                      placeholder="Add your article"
                      onChange={this.onInputChange} />
                  </FormGroup>
                </div>
                <div className="col">
                  <Button className="float-right" color="primary" size="lg">Save</Button>
                  <h2>{this.state.title}</h2>
                  <hr />
                  <ReactMarkdown source={this.state.body} />
                </div>
              </div>
            </Form>
        </div>
      </div>
    );
  }
}

const mapStateToProps = (state, props) => {
	return {
		article: state.articles.show
	};
};

export default connect(mapStateToProps, null)(ArticleEditPage);
