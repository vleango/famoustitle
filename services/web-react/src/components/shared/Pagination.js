import React, { Fragment } from 'react';
import { Link } from 'react-router-dom';
import { forEach, range } from 'lodash';
import { Pagination, PaginationItem, PaginationLink } from 'reactstrap';

import './css/pagination.css'

export default (props) => {
  const {currentPage, totalPages} = props;
  if(currentPage === 0 || currentPage > totalPages) {
    return (
      <Fragment>{''}</Fragment>
    );
  }

  const pageBoxes = 5;
  let selectedPagePosition = currentPage % pageBoxes;

  if(selectedPagePosition === 0) {
    selectedPagePosition = pageBoxes;
  }

  let paginationItems = [];
  forEach(range(1, pageBoxes + 1), (index) => {
    let pageNumber = currentPage - selectedPagePosition + index;
    if (pageNumber <= totalPages) {
      paginationItems.push(
        <PaginationItem key={"pagination-item-" + index} active={ selectedPagePosition === index }>
          <PaginationLink tag={Link} to={`/?currentPage=${pageNumber}`}>
            { pageNumber }
          </PaginationLink>
        </PaginationItem>
      );
    }
  });

  const nextPageSet = currentPage + (pageBoxes - selectedPagePosition) + 1;

  return (
    <Pagination>
      <PaginationItem hidden={currentPage <= pageBoxes }>
        <PaginationLink previous tag={Link} to={`/?currentPage=${currentPage - selectedPagePosition}`} />
      </PaginationItem>
      { paginationItems }
      <PaginationItem hidden={nextPageSet >= totalPages}>
        <PaginationLink next tag={Link} to={`/?currentPage=${nextPageSet}`} />
      </PaginationItem>
    </Pagination>
  );
}
