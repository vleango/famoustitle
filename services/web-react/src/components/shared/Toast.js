import React  from 'react';
import { ToastContainer, toast } from 'react-toastify';
import 'react-toastify/dist/ReactToastify.css';

export const toastInProgress = (message) => {
    return toast(message, { autoClose: false });
};

export const toastSuccess = (message, toastID = null) => {
    if(toastID !== null) {
        toast.update(toastID, { render: message, type: toast.TYPE.SUCCESS, autoClose: 3000 });
    } else {
        toast.success(message);
    }
};

export const toastFail = (message, toastID = null) => {
    if(toastID !== null) {
        toast.update(toastID, { render: message, type: toast.TYPE.ERROR, autoClose: 3000 });
    } else {
        toast.error(message);
    }
};

export default () => {

    return (
        <ToastContainer position="top-center"
                        autoClose={5000}
                        hideProgressBar
                        newestOnTop={false}
                        closeOnClick
                        rtl={false}
                        pauseOnVisibilityChange
                        draggable
                        pauseOnHover />
    );
}
