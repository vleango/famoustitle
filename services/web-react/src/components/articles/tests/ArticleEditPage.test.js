import React from 'react';
import { shallow } from 'enzyme';
import { ArticleEditPage } from '../ArticleEditPage';

import { toastInProgress, toastSuccess, toastFail } from '../../shared/Toast';

jest.mock('../../shared/Toast', () => ({
    toastInProgress: jest.fn((message, id) => { return 1 }),
    toastSuccess: jest.fn((message, id) => { return 2 }),
    toastFail: jest.fn((message, id) => { return 3 })
}));

beforeEach(() => {
    jest.clearAllMocks();
});

describe('Components', () => {
    describe('ArticleEditPage', () => {

        let wrapper = shallow(<ArticleEditPage />);

        describe('Snapshot', () => {
            it('should correctly render ArticleEditPage', () => {
                expect(wrapper).toMatchSnapshot();
            });
        });

        describe('onSubmitEditArticle', () => {
            it('Success', async () => {
                const updateItem = jest.fn(() => Promise.resolve());
                let wrapper = shallow(<ArticleEditPage updateItem={updateItem} history={[]} match={{params: {id: 1}}} />);
                wrapper.setState({ author: 'Tha', title: 'my title', subtitle: 'my sub', body: 'my body', tags: 'tag1,tag2' });
                await wrapper.instance().onSubmitEditArticle();
                expect(wrapper.state('submitting')).toBe(false);
                expect(toastInProgress).toHaveBeenCalledWith("Saving in progress...");
                expect(toastSuccess).toHaveBeenCalledWith("Success!", 1);
                expect(toastFail).not.toHaveBeenCalled();
            });

            it('Fail', async () => {
                const updateItem = jest.fn(() => Promise.reject());
                let wrapper = shallow(<ArticleEditPage updateItem={updateItem} history={[]} match={{params: {id: 1}}} />);
                wrapper.setState({ author: 'Tha', title: 'my title', subtitle: 'my sub', body: 'my body', tags: 'tag1,tag2' });
                await wrapper.instance().onSubmitEditArticle();
                expect(wrapper.state('submitting')).toBe(false);
                expect(wrapper.state('errorMsg')).toBe("server error");
                expect(toastInProgress).toHaveBeenCalledWith("Saving in progress...");
                expect(toastSuccess).not.toHaveBeenCalled();
                expect(toastFail).toHaveBeenCalledWith("server error", 1);
            });
        });
    });
});
