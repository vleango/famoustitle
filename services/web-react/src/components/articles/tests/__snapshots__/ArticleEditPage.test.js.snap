// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Components ArticleEditPage Snapshot should correctly render ArticleEditPage 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <ArticleEditPage />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": Array [
        <HelmetWrapper
          defer={true}
          encodeSpecialCharacters={true}
>
          <title>
                    Edit an article - FamousTitle.com
          </title>
</HelmetWrapper>,
        <div
          className="container p-5"
>
          <Unknown />
          
</div>,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": <title>
            Edit an article - FamousTitle.com
</title>,
          "defer": true,
          "encodeSpecialCharacters": true,
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Edit an article - FamousTitle.com",
          },
          "ref": null,
          "rendered": "Edit an article - FamousTitle.com",
          "type": "title",
        },
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <Unknown />,
            "",
          ],
          "className": "container p-5",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {},
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          "",
        ],
        "type": "div",
      },
    ],
    "type": Symbol(react.fragment),
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": Array [
          <HelmetWrapper
            defer={true}
            encodeSpecialCharacters={true}
>
            <title>
                        Edit an article - FamousTitle.com
            </title>
</HelmetWrapper>,
          <div
            className="container p-5"
>
            <Unknown />
            
</div>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": <title>
              Edit an article - FamousTitle.com
</title>,
            "defer": true,
            "encodeSpecialCharacters": true,
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Edit an article - FamousTitle.com",
            },
            "ref": null,
            "rendered": "Edit an article - FamousTitle.com",
            "type": "title",
          },
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <Unknown />,
              "",
            ],
            "className": "container p-5",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {},
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            "",
          ],
          "type": "div",
        },
      ],
      "type": Symbol(react.fragment),
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
